wis2box/oregon/main.py

```diff
@@ -0,0 +1,298 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+###############################################################################
+
+import json
+import requests
+import io
+from typing import Tuple, TypedDict, Optional, List
+import click
+import logging
+from wis2box import cli_helpers, data
+from wis2box.api import remove_collection, setup_collection, upsert_collection_item
+from urllib.parse import parse_qs, urlencode, urlparse
+from requests import Session
+
+from wis2box.api.backend import load_backend
+from wis2box.api.backend.sensorthings import SensorthingsBackend
+from wis2box.env import API_BACKEND_URL, STORAGE_INCOMING
+from wis2box.oregon.cache import ShelveCache
+from wis2box.oregon.lib import OregonHttpClient, download_oregon_tsv, generate_phenomenon_time, parse_date, parse_oregon_tsv
+from wis2box.oregon.types import ALL_RELEVANT_STATIONS, POTENTIAL_DATASTREAMS, Attributes, OregonHttpResponse, StationData, Datastream
+from wis2box.storage import put_data
+
+LOGGER = logging.getLogger(__name__)
+
+THINGS_COLLECTION = "Things"
+
+
+class OregonStaRequestBuilder():
+
+    relevant_stations: List[int]
+
+    def __init__(self, relevant_stations: List[int]) -> None:
+        self.relevant_stations = relevant_stations
+
+    def _get_upstream_data(self) -> list[StationData]:
+        """Get metadata for all relevant stations."""
+        client = OregonHttpClient()
+        # Split the ALL_RELEVANT_STATIONS into two halves since the oregon api can't handle all of them at once
+        half_index = len(self.relevant_stations) // 2
+        first_half_stations = self.relevant_stations[:half_index]
+        second_half_stations = self.relevant_stations[half_index:]
+
+        # Fetch and process the first half of the stations
+        first_station_set: OregonHttpResponse = client.fetch_stations(first_half_stations)
+        second_station_set: OregonHttpResponse = client.fetch_stations(second_half_stations)
+        # create one big dict
+
+        all_stations = first_station_set["features"] + second_station_set["features"]
+        return all_stations
+
+    def _get_data_associated_with_station(
+        self, station_nbr, start_date, end_date, dataset
+    ) -> Tuple[list[float], list[str]]:
+        """Ingest all data from a station and return the third column."""
+        start_date = "09/25/2024 12:00:00"
+        end_date = "09/30/2024 12:00:00"
```

Comment:
Currently hard coded for right now since I don't want us accidentally loading in all the data and spamming them